packages/devtools-proxy-support/src/socks5.ts

```diff
@@ -0,0 +1,382 @@
+
+    const listeningPromise = this.serverListen(proxyPort, proxyHost);
+    try {
+      await Promise.all([
+        listeningPromise,
+        once(this, 'listening'),
+        this.ensureAgentInitialized(),
+      ]);
+      this.agentInitialized = true;
+    } catch (err: unknown) {
+      await this.close();
+      throw err;
+    }
+  }
+
+  private async ensureAgentInitialized() {
+    if (this.agentInitialized) {
+      return;
+    }
+
+    if (this.agentInitPromise) {
+      return this.agentInitPromise;
+    }
+
+    if (this.closed) {
+      // A socks5 request could come in after we deliberately closed the connection. Don't reconnect in that case.
+      throw new Error('Disconnected.');
+    }
+
+    try {
+      await (this.agentInitPromise = this.agent.initialize?.());
+    } catch (err) {
+      this.emit('forwardingError', err);
+      this.logger.emit('socks5:forwarding-error', {
+        error: (err as any)?.stack ?? String(err),
+      });
+      delete this.agentInitPromise;
+      await this.serverClose();
+      throw err;
+    }
+
+    delete this.agentInitPromise;
+    this.agentInitialized = true;
+    this.logger.emit('socks5:agent-initialized');
+  }
+
+  private async closeOpenConnections() {
+    const waitForClose: Promise<unknown[]>[] = [];
+    for (const socket of this.connections) {
+      waitForClose.push(once(socket, 'close'));
+      socket.destroy();
+    }
+    await Promise.all(waitForClose);
+    this.connections.clear();
+  }
+
+  async close(): Promise<void> {
+    if (this.closed) return;
+    this.closed = true;
+
+    this.logger.emit('socks5:closing-tunnel');
+    const [maybeError] = await Promise.all([
+      // If we catch anything, just return the error instead of throwing, we
+      // want to await on closing the connections before re-throwing server
+      // close error
+      this.serverClose().catch<Error>((e) => e),
+      this.agent.destroy?.(),
+      this.closeOpenConnections(),
+    ]);
+
+    if (
+      maybeError &&
+      !('code' in maybeError && maybeError.code === 'ERR_SERVER_NOT_RUNNING')
+    ) {
+      throw maybeError;
+    }
+  }
+
+  private async forwardOut(dstAddr: string, dstPort: number): Promise<Duplex> {
+    const channel = await new Promise<Duplex>((resolve, reject) => {
+      const req = createFakeHttpClientRequest(
+        dstAddr,
+        dstPort,
+        this.overrideProtocol
+      );
+      req.onSocket = (sock) => {
+        if (sock) resolve(sock);
+      };
+      this.agent.createSocket(
+        req,
+        {
+          host: dstAddr,
+          port: dstPort,
+        },
+        (err, sock) => {
+          // Ideally, we would always be using this callback for retrieving the `sock`
+          // instance. However, agent-base does not call the callback at all if
+          // the agent resolved to another agent (as is the case for e.g. `ProxyAgent`).
+          if (err) reject(err);
+          else if (sock) resolve(sock);
```

Comment:
What if there is no error but sock is undefined? this promise remains unresolved?

Reply:
If neither this line nor L247 are triggered with a valid socket, then yeah, the Promise would remain unresolved.

Tbh, I think I did this to appease typescript – either `err` or `sock` should always be truth-y here

Reply:
c6d99f8654f8ee3bc407272c91c62c47633b69b4 should help with this?